jena-arq/src/main/java/org/apache/jena/sparql/engine/main/OpExecutor.java

```diff
@@ -210,10 +211,19 @@ protected QueryIterator execute(OpJoin opJoin, QueryIterator input) {
             QueryIterator qIter = Join.join(left, right, execCxt);
             return qIter;
         }
+
+        JoinKey joinKey = null;
+        if (false) {
+            // Disabled because variable analysis may fail when custom ops are involved.
+            Set<Var> leftVars = OpVars.visibleVars(opJoin.getLeft());
+            Set<Var> rightVars = OpVars.visibleVars(opJoin.getRight());
+            joinKey = JoinKey.create(leftVars, rightVars);
+        }
+
         QueryIterator left = exec(opJoin.getLeft(), input);
         QueryIterator right = exec(opJoin.getRight(), root());
-        // Join key.
-        QueryIterator qIter = Join.join(left, right, execCxt);
+
+        QueryIterator qIter = Join.join(joinKey, left, right, execCxt);
```

Comment:
I'm confused. `joinKey` is always null.

I think joinKey is not relevant here and the call should be `Join.join(left, right, execCxt)`, then `Join.join` calculates the join key, making join keys internal to the join code.

Reply:
Removed joinKey argument.